macaroonbakery/macaroon.py

```diff
@@ -0,0 +1,285 @@
+# Copyright 2017 Canonical Ltd.
+# Licensed under the LGPLv3, see LICENCE file for details.
+
+import base64
+import copy
+import log
+import os
+
+import bakery
+import codec
+import pymacaroons
+
+MACAROON_V1, MACAROON_V2 = 1, 2
+
+
+class Macaroon:
+    '''Represent an undischarged macaroon along its first
+    party caveat namespace and associated third party caveat information
+    which should be passed to the third party when discharging a caveat.
+    '''
+    def __init__(self, root_key, id, location=None,
+                 version=bakery.LATEST_BAKERY_VERSION, ns=None):
+        '''Create a new macaroon with the given root key, id and location.
+
+        If the version is more than the latest known version,
+        the latest known version will be used. The namespace is that
+        of the service creating it.
+        @param root_key bytes or string
+        @param id bytes or string
+        @param location bytes or string
+        @param version the bakery version.
+        @param ns
+        '''
+        if version > bakery.LATEST_BAKERY_VERSION:
+            log.info('use last know version:{} instead of: {}'.format(
+                bakery.LATEST_BAKERY_VERSION, version
+            ))
+            version = bakery.LATEST_BAKERY_VERSION
+        # m holds the underlying macaroon.
+        self._macaroon = pymacaroons.Macaroon(location=location, key=root_key,
+                                              identifier=id)
+        # version holds the version of the macaroon.
+        self.version = version
+        self.caveat_data = {}
+
+    def add_caveat(self, cav, key=None, loc=None):
+        '''Return a new macaroon with the given caveat.
+
+        It encrypts it using the given key pair
+        and by looking up the location using the given locator.
+        As a special case, if the caveat's Location field has the prefix
+        "local " the caveat is added as a client self-discharge caveat using
+        the public key base64-encoded in the rest of the location. In this
+        case, the Condition field must be empty. The resulting third-party
+        caveat will encode the condition "true" encrypted with that public
+        key.
+
+        @param cav the caveat to be added.
+        @param key the nacl public key to encrypt third party caveat.
+        @param loc locator to find the location object that has a method
```

Comment:
@param loc locator to find information on third parties when adding
third party caveats. It is expected to have a third_party_info method
that will be called with a location string and should return a ThirdPartyInfo
instance holding the requested information.

?

(what's the expected behaviour when third_party_info cannot find information
on the third party?)